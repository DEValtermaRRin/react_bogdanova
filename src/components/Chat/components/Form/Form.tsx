import React, { useState, memo } from 'react';
import { Button } from './components/Button/Button';
import { Message } from './components/Message/Message';
import './Form.scss';

interface FormProps {
  addMessages: (val: string, user: string) => void;
  onSubmit?: () => void;
  userName: string;
}

export const Form = memo<FormProps>(({ userName, addMessages }) => {
  const name = 'Send message';
  const [value, setValue] = useState('');

  const handleClickSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    addMessages(value, userName);
    setValue('');
  };
  return (
    <form data-testid="form" onSubmit={handleClickSubmit} className="form">
      <Message value={value} setValue={setValue} />
      <Button disabled={!userName} name={name} />
    </form>
  );
});
