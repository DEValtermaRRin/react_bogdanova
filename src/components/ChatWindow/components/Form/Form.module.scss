.form {
  margin-top: 5px;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  height: 40px;
  width: 560px;
  &__flex {
    display: flex;
    align-items: center;
  }
}

.form__flexAround {
  display: flex;
  justify-content: space-around;
  &__flex {
    display: flex;
    align-items: center;
  }
}