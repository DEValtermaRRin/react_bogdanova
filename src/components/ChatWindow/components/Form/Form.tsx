import React, { useState, memo } from 'react';
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';

import { ButtonSend } from './components/ButtonSend';
import { ButtonDel } from './components/ButtonDel';
import { Message } from './components/Message';

import { addMessage } from 'src/store/chatlist/slice';

import style from './Form.module.scss';

export interface FormProps {
  userName: string;
}

export const Form = memo<FormProps>(({ userName }) => {
  const [value, setValue] = useState('');
  const dispatch = useDispatch();
  const { chatId } = useParams();

  const handleClickSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (chatId && value) {
      dispatch(
        addMessage({ chatId, message: { text: value, author: userName } }),
      );
    }
    setValue('');
  };
  return (
    <form
      data-testid="form"
      onSubmit={handleClickSubmit}
      className={style.form}
    >
      <ButtonDel />
      <Message value={value} setValue={setValue} />
      <ButtonSend disabled={!userName} />
    </form>
  );
});
