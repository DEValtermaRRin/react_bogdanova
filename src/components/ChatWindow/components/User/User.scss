.user {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 30px;
  height: 30px;
  background-color: #284779;
  overflow: hidden;
  border-radius: 15px;
  margin: 3px;
  &:hover {
    animation: user-dragout 0.5s ease-in-out 0.1s 1 normal forwards;
  }

  &:hover &__input {
    animation: input-dragout 0.5s ease-in-out 0.1s 1 normal forwards;
  }

  &__icon {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 30px;
    height: 30px;
    background: url(./assets/user.svg) 100%;
    background-color: #284779;
    border-radius: 50%;
    overflow: hidden;
    &:hover {
      cursor: pointer;
    }
  }
  &__input {
    height: 26px;
    border-radius: 15px;
    border: none;
    background: #252424;
    font-size: 14px;
    font-weight: 600;
    letter-spacing: 0.1rem;
    color: #366dc5;
    opacity: 0;
    width: 0;
    margin-right: 0px;
    &::placeholder {
      color: #4979af75;
    }
  }
}

@keyframes user-dragout {
  0% {
    width: 35px;
  }
  100% {
    width: 250px;
  }
}

@keyframes input-dragout {
  0% {
    padding: 0;
    opacity: 0;
    width: 0;
    margin-right: 0px;
  }
  100% {
    padding: 3px;
    opacity: 1;
    width: 200px;
    margin-right: 2px;
  }
}
