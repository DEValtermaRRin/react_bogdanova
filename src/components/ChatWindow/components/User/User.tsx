import React, { FC } from 'react';
import style from './User.module.scss';

interface UserProps {
  name: string;
  getName: (e: string) => void;
}

export const User: FC<UserProps> = ({ name, getName }) => (
  <div className={style.user}>
    <div className={style.user__icon}></div>
    <input
      className={style.user__icon}
      data-testid="user"
      value={name}
      onChange={(e) => getName(e.target.value)}
      type="text"
      placeholder="Enter your name"
    />
  </div>

  
);
