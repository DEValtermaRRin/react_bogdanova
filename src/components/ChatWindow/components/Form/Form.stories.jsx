import React from 'react';
import { Form } from './Form';

export default {
  component: Form,
};

const Template = (args) => <Form {...args} />;

export const Test = Template.bind({});
