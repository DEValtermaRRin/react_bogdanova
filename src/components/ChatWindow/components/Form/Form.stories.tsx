import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Form, FormProps } from './Form';

import style from './Form.module.scss';

export default {
  title: 'Components/ChatWindow/Components/Form',
  component: Form,
  argTypes: {
    variant: {
      type: 'string',
      description: 'Вариант расположения элементов в форме',
      defaultValue: style.form,
      options: [style.form__flexAround],
      control: {
        type: 'radio',
      },
    },
  },
} as ComponentMeta<typeof Form>;

const Template: ComponentStory<typeof Form> = (args: FormProps) => (
  <Form {...args} />
);

export const Default = Template.bind({});

Default.arguments = {
  variant: style.form,
};
