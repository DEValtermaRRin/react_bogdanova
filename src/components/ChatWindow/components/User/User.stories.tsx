import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { User, UserProps } from './User';
import style from './User.module.scss';

export default {
  title: 'Components/ChatWindow',
  component: User,
  argTypes: {
    variant: {
      type: 'string',
      description: 'Выплывающий инпут',
      defaultValue: style.user,
      control: { background: 'color' },
    },
  },
} as ComponentMeta<typeof User>;

const Template: ComponentStory<typeof User> = (args: UserProps) => (
  <User {...args} />
);

export const Default = Template.bind({});

Default.arguments = { variant: style.user };
