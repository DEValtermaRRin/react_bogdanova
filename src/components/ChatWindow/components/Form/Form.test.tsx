import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Form } from './Form';
import { User } from '../User/User';

describe('Form', () => {
  it('render component', () => {
    render(<Form userName={''} />);
  });
  it('render with snapshot', () => {
    expect(render(<Form userName={''} />)).toMatchSnapshot();
  });
  it('render message with text', () => {
    render(<Form userName={''} />);
    fireEvent.input(screen.getByTestId('message'), {
      target: { value: 'Hello' },
    });
    fireEvent.click(screen.getByTestId('button'));
    expect(screen.getByText(/hello/i)).toBeInTheDocument();
  });
  it('check click', () => {
    const handleSubmit = jest.fn();
    render(<User name="Vasia" getName={handleSubmit} />);
    const { getByTestId } = render(<Form userName={''} />);
    const button = getByTestId('button');
    fireEvent.input(screen.getByTestId('user'), {
      target: { value: 'Vasya' },
    });
    fireEvent.click(button);
    expect(handleSubmit).toHaveBeenCalledTimes(1);
  });
});
