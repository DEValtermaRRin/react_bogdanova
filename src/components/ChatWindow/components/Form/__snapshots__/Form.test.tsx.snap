// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="form"
      >
        <button
          class="delbutton"
          type="button"
        >
          <svg
            viewBox="0 0 96 96"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <path
              d="m24,78c0,4.968 4.029,9 9,9h30c4.968,0 9-4.032 9-9l6-48h-60l6,48zm33-39h6v39h-6v-39zm-12,0h6v39h-6v-39zm-12,0h6v39h-6v-39zm43.5-21h-19.5c0,0-1.344-6-3-6h-12c-1.659,0-3,6-3,6h-19.5c-2.487,0-4.5,2.013-4.5,4.5s0,4.5 0,4.5h66c0,0 0-2.013 0-4.5s-2.016-4.5-4.5-4.5z"
            />
          </svg>
        </button>
        <textarea
          class="message"
          data-testid="message"
          placeholder="Your message"
        />
        <button
          class="btn"
          data-testid="button"
          disabled=""
          type="submit"
        >
          <svg
            class="ionicon"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Enter your name
            </title>
            <path
              d="M470.3 271.15L43.16 447.31a7.83 7.83 0 01-11.16-7V327a8 8 0 016.51-7.86l247.62-47c17.36-3.29 17.36-28.15 0-31.44l-247.63-47a8 8 0 01-6.5-7.85V72.59c0-5.74 5.88-10.26 11.16-8L470.3 241.76a16 16 0 010 29.39z"
            />
          </svg>
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="form"
    >
      <button
        class="delbutton"
        type="button"
      >
        <svg
          viewBox="0 0 96 96"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <path
            d="m24,78c0,4.968 4.029,9 9,9h30c4.968,0 9-4.032 9-9l6-48h-60l6,48zm33-39h6v39h-6v-39zm-12,0h6v39h-6v-39zm-12,0h6v39h-6v-39zm43.5-21h-19.5c0,0-1.344-6-3-6h-12c-1.659,0-3,6-3,6h-19.5c-2.487,0-4.5,2.013-4.5,4.5s0,4.5 0,4.5h66c0,0 0-2.013 0-4.5s-2.016-4.5-4.5-4.5z"
          />
        </svg>
      </button>
      <textarea
        class="message"
        data-testid="message"
        placeholder="Your message"
      />
      <button
        class="btn"
        data-testid="button"
        disabled=""
        type="submit"
      >
        <svg
          class="ionicon"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Enter your name
          </title>
          <path
            d="M470.3 271.15L43.16 447.31a7.83 7.83 0 01-11.16-7V327a8 8 0 016.51-7.86l247.62-47c17.36-3.29 17.36-28.15 0-31.44l-247.63-47a8 8 0 01-6.5-7.85V72.59c0-5.74 5.88-10.26 11.16-8L470.3 241.76a16 16 0 010 29.39z"
          />
        </svg>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
